_posts/2017-03-30-storage-formats-quickstep.markdown

```diff
@@ -0,0 +1,140 @@
+
+In the row store format, a tuple is stored in the table definition order. In memory, the ID (101) of the first tuple is stored at some address, and it is followed by the age field for that tuple (30), and then the name field ("Jennifer"). The second tuple is stored in memory sequentially after the first tuple, and so on.
+
+<h2>Column Store</h2>
+
+Let's now take a look at the *column store* format. The figure below depicts a column store layout for our toy example. 
+
+![Column Store](../assets/storage-format-column-store.jpg)
+
+In the column store format, the values from the same column are stored together. The order in which the values from a column are stored remains the same for all the columns. Notice the ID values 101, 102, 103 are stored contiguously. The corresponding Age fields 30, 25 and 35 are stored contiguously and likewise for the Name column.
+
+<h2>Compression</h2>
+
+Compression is a standard technique to reduce the storage footprint. There is a large body of work on various compression techniques, which we won't cover here. Let's look at how compression can be applied in our toy example. If we look at the ID column, it has three values 101, 102 and 103. If they are stored as regular integers, each of them will occupy 4 bytes of memory. Interestingly, we can reduce the memory consumed in storing these three values. Observe that these three values are very close to 100. If we remember the base value 100 and only record the difference of each value from this base value (of 100), then we only need to store the values 1, 2, and 3. To store these three *deltas*, we don't even need a 4 byte integer, a 1-byte value is enough.
+
+Putting it all together, the picture below shows the compressed column store format in which the compression is applied on the ID column. 
+
+![Column Store with Compression](../assets/storage-format-compressed-column-store.jpg)
+
+Observe that the memory occupied by the three tuples together is only 43 bytes, compared to 48 bytes in the row store and the column store formats. As the number of tuples increase, there may be more opportunities for compression and thus more memory savings.
+
+There are some aspects related to storage management which deserve a blog post of their own. This includes the block-based storage design, and the impact of the above storage formats on various analytical queries. I hope to write more blog posts to cover these topics in the future.
+
+<h1>Creating tables with various storage formats in Quickstep</h1>
+
+Now I will show you how to use the above storage formats to create tables in Quickstep. The storage format specification is part of the `CREATE TABLE SQL` command. We will continue with our toy example above to illustrate the various ways in which we can use the storage formats in Quickstep. 
+
+{% highlight sql %}
+CREATE TABLE employee (
+ID INT NOT NULL, 
+Age INT NOT NULL, 
+Name VARCHAR(8) NOT NULL
+) WITH BLOCKPROPERTIES (
+  TYPE rowstore,
```

Comment:
looks like 'rowstore' doesn't work. I think we should actually change quickstep to support the keyworkd again, because it makes more sense than `split_rowstore` (which does work)